### PR TITLE
refactor(laws): autocomplete multiselect for laws advanced search

### PR DESCRIPTION
**Pull request by juleskuehn**

### CODE REVIEW

File: django/chat/llm.py [hunk 1]
```diff
@@ -43,6 +47,7 @@ def __init__(
         )
         self._callback_manager = CallbackManager([self._token_counter])
         self.llm = self._get_llm()
+        self.mock_embedding = mock_embedding
```
Comment (juleskuehn) [inline django/chat/llm.py:50]: The Mock embedding is for costing or testing purposes. e.g. skipping embedding a bunch of a laws when I'm trying to test how they are loaded. We should add a Mock LLM here later as well, for estimating costs before running expensive processes.

File: django/laws/forms.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                {"label": str(x), "value": str(x.id)}
+                for x in data
+                if str(x.id) in values
+            ]
+            return items
+
+        return []
+
+
+class LawSearchForm(forms.Form):
```
Comment (juleskuehn) [inline django/laws/forms.py:69]: The main work of this PR is the new advanced search form.

File: django/laws/management/commands/load_laws_xml.py [hunk 3]
```diff
@@ -680,180 +755,283 @@ def handle(self, *args, **options):
...
+                document_fr,
+                nodes_fr,
+                sha_256_hash_en=en_hash,
+                sha_256_hash_fr=fr_hash,
+                add_to_vector_store=True,
+                force_update=force_update,
+                llm=llm,
+            )
+            bind_contextvars(law_id=law.id)
+            llm.create_costs()
```
Comment (juleskuehn) [inline django/laws/management/commands/load_laws_xml.py:1029]: Costs for embedding laws are saved and associated with particular Law object.

File: django/laws/views.py [hunk 4]
```diff
@@ -82,34 +86,42 @@ def _get_other_lang_node(node_id):
     return _get_source_node(other_lang_node_id)
 
 
-def _get_law_url(law):
+def _get_law_url(law, request_lang):
```
Comment (juleskuehn) [inline django/laws/views.py:89]: Law objects are now Bilingual so some refactoring was required.

File: django/laws/views.py [hunk 5]
```diff
@@ -372,19 +324,23 @@ def htmx_sse_response(response_gen):
             message_html = llm_response_to_html(full_message)
             message_html_lines = message_html.split("\n")
 
+        sync_to_async(llm.create_costs)()
```
Comment (juleskuehn) [inline django/laws/views.py:327]: The query costing isn't working yet. Not sure why.

File: django/laws/views.py [hunk 6]
```diff
@@ -372,19 +324,23 @@ def htmx_sse_response(response_gen):
...
+        sync_to_async(cache.delete)(f"sources_{query}")
+
         yield (
             f"data: <div hx-swap-oob='true' id='answer-sse'>"
             f"<div>{sse_joiner.join(message_html_lines)}</div></div>\n\n"
```
Comment (juleskuehn) [inline django/laws/views.py:332]: Sometimes the AI answer gets streamed out, then disappears and starts over, then completes. Not sure why. This problem may predate this PR.

File: django/requirements.txt [hunk 7]
```diff
@@ -18,6 +18,7 @@ rules==3.4
 django-cleanup==8.1.0
 django-azure-auth==2.0.0
 gevent==24.2.1
+django-data-fetcher==2.2
```
Comment (juleskuehn) [inline django/requirements.txt:21]: Not used in this PR so I could remove, but will need to use it very shortly for another ticket.

**Comment by smcdonald-jus** [comment]
The old form had a few options for searching across enabling authorities that have been removed here (the ones circled below). Are these options worth preserving?

image

**Comment by juleskuehn** [comment]
> The old form had a few options for searching across enabling authorities that have been removed here (the ones circled below). Are these options worth preserving?
> 
> image

Indeed. Nobody has really asked for these, and there is a work around to get those functions so I think it is OK to remove them for now. (For "enabling authority of selected": Just find the document you want with basic search, click kebab icon, you'll see the enabling authority. Then search for that using "select acts/regs"; For "Same enabling authority", again, find the enabling authority you want, then search for that using "enabling authority")

**Review by haitmo — APPROVED** [review]

**Pull request merged**